### ISSUE TITLE
Integration: end-to-end workflow tests across the full ecosystem

### ISSUE THREAD

**Issue by nster101** [nster101/nemopy#67]
## Summary

Create an end-to-end integration test that exercises a realistic data science workflow entirely through `import nemopy as np`, combining multiple ecosystem libraries (pandas/polars data loading → nemopy computation → scipy optimization → scikit-learn modeling → matplotlib visualization). This is the capstone test that proves the drop-in replacement story works in practice.

## Why this matters

Individual library compat tests (matplotlib, scipy, pandas, polars, sklearn) verify pairwise interactions. But real code chains multiple libraries together in a single script. A bug that only surfaces when data flows through the full pipeline — e.g., a type that survives scipy but crashes matplotlib, or a shape that works in sklearn but breaks when fed back into a nemopy operation — won't be caught by isolated tests.

This issue is about the **end-to-end workflow**, not individual library compatibility.

### Step 1: Create test file

Create `tests/test_integration_workflow.py`.

### Step 2: Workflow scenarios

**Workflow 1: OLS regression — data to plot (the canonical workflow)**

```python
import nemopy as np
from nemopy import _c, mat, as_col, as_mat, ColVec, Mat

def test_ols_end_to_end():
    """
    Full OLS workflow:
    1. Create data (or load from pandas/polars)
    2. Build design matrix with nemopy
    3. Compute OLS coefficients
    4. Predict
    5. Compute residuals
    6. Plot results with matplotlib
    """
    import pandas as pd
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    # Step 1: Data from pandas
    df = pd.DataFrame({
        "x": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
        "y": [2.1, 3.9, 6.2, 7.8, 10.1, 12.0, 13.9, 16.2, 17.8, 20.1],
    })

    x = as_col(df["x"])
    y = as_col(df["y"])

    # Step 2: Design matrix with intercept
    ones = _c[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
    X = ones | x  # (10, 2) Mat — intercept + slope

    # Step 3: OLS beta = (X'X)^-1 X'y
    beta = (X.T @ X).inv @ X.T @ y
    assert isinstance(beta, ColVec)
    assert beta.shape == (2, 1)

    # Step 4: Predictions
    y_hat = X @ beta
    assert isinstance(y_hat, ColVec)
    assert y_hat.shape == (10, 1)

    # Step 5: Residuals
    residuals = y - y_hat
    assert isinstance(residuals, ColVec)
    sse = (residuals.T @ residuals).item()
    assert sse < 1.0  # good fit

    # Step 6: Plot
    fig, axes = plt.subplots(1, 2, figsize=(10, 4))
    axes[0].scatter(x.to_flat(), y.to_flat(), label="data")
    axes[0].plot(x.to_flat(), y_hat.to_flat(), 'r-', label="fit")
    axes[0].set_title("OLS Fit")
    axes[0].legend()

    axes[1].scatter(y_hat.to_flat(), residuals.to_flat())
    axes[1].axhline(y=0, color='r', linestyle='--')
    axes[1].set_title("Residuals")

    plt.tight_layout()
    plt.savefig("/tmp/nemopy_ols_test.png", dpi=72)
    plt.close(fig)


def test_ols_vs_sklearn():
    """
    Verify nemopy OLS matches sklearn's LinearRegression.
    Proves correctness of the manual linear algebra path.
    """
    from sklearn.linear_model import LinearRegression

    X_data = mat([1, 2, 3, 4, 5], [2, 3, 4, 5, 6])  # (5,2)
    y_data = _c[8, 13, 18, 23, 28]                    # y = 2*x1 + 3*x2

    # nemopy OLS
    ones = _c[1, 1, 1, 1, 1]
    X_design = ones | X_data                            # (5,3) with intercept
    beta_nemo = (X_design.T @ X_design).inv @ X_design.T @ y_data

    # sklearn OLS
    model = LinearRegression()
    model.fit(X_data, y_data.to_flat())

    # Compare (sklearn separates intercept from coefs)
    np.testing.assert_allclose(beta_nemo[0], model.intercept_, atol=1e-8)
    np.testing.assert_allclose(beta_nemo[1:].to_flat(), model.coef_, atol=1e-8)
```

**Workflow 2: Eigendecomposition and visualization**

```python
def test_eigen_analysis_workflow():
    """
    PCA-like workflow:
    1. Build covariance matrix from data
    2. Eigendecompose with scipy
    3. Project data
    4. Visualize
    """
    import scipy.linalg
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    # Correlated 2D data
    data = mat(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
        [2.1, 3.9, 6.2, 7.8, 10.1, 12.0, 13.9, 16.2, 17.8, 20.1]
    )  # (10, 2)

    # Center the data
    mean_vec = _c[
        data[:, 0].mean(),
        data[:, 1].mean()
    ]
    # This is the kind of operation that tests type persistence
    centered = data - (np.ones((10, 1)) @ mean_vec.T)

    # Covariance matrix
    n = data.shape[0]
    cov = (centered.T @ centered) * (1.0 / (n - 1))
    assert isinstance(cov, Mat)
    assert cov.shape == (2, 2)

    # Eigendecomposition
    eigenvalues, eigenvectors = scipy.linalg.eigh(cov)

    # Project onto first principal component
    pc1 = eigenvectors[:, -1].reshape(-1, 1)  # largest eigenvalue
    projected = centered @ pc1
    assert projected.shape[0] == 10

    # Plot
    fig, ax = plt.subplots()
    ax.scatter(data[:, 0].flatten(), data[:, 1].flatten())
    ax.set_title("Eigenanalysis")
    plt.savefig("/tmp/nemopy_eigen_test.png", dpi=72)
    plt.close(fig)
```

**Workflow 3: scipy.optimize with nemopy objective**

```python
def test_optimization_workflow():
    """
    Optimization:
    1. Define objective using nemopy linear algebra
    2. Optimize with scipy
    3. Verify solution
    """
    import scipy.optimize

    A = mat([2, 1], [1, 3])
    b = _c[5, 7]

    # Objective: minimize ||Ax - b||^2
    def objective(x_flat):
        x = as_col(x_flat)
        r = A @ x - b
        return float((r.T @ r).item())

    x0 = np.zeros(2)
    result = scipy.optimize.minimize(objective, x0, method='BFGS')
    assert result.success

    # Verify: solution should equal A^{-1} b
    x_opt = as_col(result.x)
    x_exact = A.inv @ b
    np.testing.assert_allclose(x_opt.to_flat(), x_exact.to_flat(), atol=1e-6)
```

**Workflow 4: Polars → nemopy → sklearn → matplotlib**

```python
def test_polars_to_sklearn_to_plot():
    """
    Full pipeline:
    1. Load from Polars DataFrame
    2. Extract features/target with nemopy
    3. Train sklearn model
    4. Plot predictions with matplotlib
    """
    import polars as pl
    import nemopy.polars
    from sklearn.linear_model import LinearRegression
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    # Step 1: Polars data
    df = pl.DataFrame({
        "sq_ft": [800, 1000, 1200, 1400, 1600, 1800, 2000],
        "price": [150, 200, 240, 280, 330, 365, 400],
    })

    # Step 2: Extract with nemopy
    X = df.nemo.mat(["sq_ft"])
    y = df.nemo.col("price")

    # Step 3: Train
    model = LinearRegression()
    model.fit(X, y.to_flat())
    preds = model.predict(X)

    # Step 4: Plot
    fig, ax = plt.subplots()
    ax.scatter(X.to_flat() if hasattr(X, 'to_flat') else X.flatten(),
               y.to_flat(), label="actual")
    ax.plot(X.to_flat() if hasattr(X, 'to_flat') else X.flatten(),
            preds, 'r-', label="predicted")
    ax.set_xlabel("sq_ft")
    ax.set_ylabel("price")
    ax.legend()
    plt.savefig("/tmp/nemopy_pipeline_test.png", dpi=72)
    plt.close(fig)
```

### Step 3: Failure triage

Any failure in these workflows indicates a real interop gap that would block users. Categorize as:

| Type | Example | Action |
|------|---------|--------|
| **Type loss** | `X @ beta` returns ndarray not ColVec | Fix in `__array_function__` or `__array_ufunc__` |
| **Shape mismatch** | Library expects 1D, gets `(n,1)` | Document `.to_flat()` pattern; consider if nemopy should auto-flatten in certain contexts |
| **Assertion error** | Values wrong | Bug in nemopy or test — investigate |
| **Import error** | Namespace passthrough gap | Fix in `__init__.py` |

### Files to create
- `tests/test_integration_workflow.py`

### Dependencies on other issues
- Namespace passthrough (#51) — must be complete
- `__array_function__` (#50) — must be complete
- Individual compat issues (matplotlib, scipy, pandas, polars, sklearn) — ideally complete but not strictly required. This test will surface the same issues if they aren't.

### Reference
- All nemopy source files in `nemopy/`
- DESIGN.md §8 (linear algebra operations), §11-13 (interoperability), §19 (Polars)
- nemopy `ColVec.to_flat()`, `as_col()`, `as_mat()` for conversion patterns
